pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -875,11 +875,8 @@ protected WriteInEventLoopCallback newObject(Handle<WriteInEventLoopCallback> ha
         ClientCnx cnx = cnx();
         if (cnx == null || currentState != State.Ready) {
             log.info("[{}] [{}] Closed Producer (not connected)", topic, producerName);
-            synchronized (this) {
-                setState(State.Closed);
-                client.cleanupProducer(this);
-                clearPendingMessagesWhenClose();
-            }
+            client.cleanupProducer(this);
```

Comment:
Would it make sense to also move `client.cleanupProducer(this);` to  `closeAndClearPendingMessages()`?

Reply:
Yes, we can add that to ` closeAndClearPendingMessages`.

Reply:
I put it in the sync block based on the same reasoning from the discussion on `setState`.